File: frontends/react-ts/src/stores/capsule.ts
// ============================================================
// 胶囊状态：我创建的 + 我收藏的（分页 / 删除 / 收藏切换）
// ============================================================

import { create } from "zustand";
import { api } from "@/api/client";
import { usePlaza } from "@/stores/plaza";
import type { CapsuleListItem, Pagination } from "@/types";

const PAGE_SIZE = 15;

interface ListSlice {
  items: CapsuleListItem[];
  pagination: Pagination | null;
  page: number;
  loading: boolean;
  error: string | null;
}

const emptySlice = (): ListSlice => ({
  items: [],
  pagination: null,
  page: 1,
  loading: false,
  error: null,
});

interface CapsuleState {
  mine: ListSlice;
  favorites: ListSlice;

  fetchMine: (page?: number) => Promise<void>;
  fetchFavorites: (page?: number) => Promise<void>;
  deleteCapsule: (id: string) => Promise<void>;
  toggleFavorite: (
    capsuleId: string,
    favoritedByMe: boolean,
  ) => Promise<{ favoriteCount: number }>;
  reset: () => void;
}

// 序列号：避免并发请求乱序覆盖结果
let mineSeq = 0;
let favSeq = 0;

export const useCapsule = create<CapsuleState>()((set, get) => ({
  mine: emptySlice(),
  favorites: emptySlice(),

  fetchMine: async (page) => {
    const p = page ?? get().mine.page;
    const myId = ++mineSeq;
    set((s) => ({ mine: { ...s.mine, page: p, loading: true, error: null } }));
    try {
      const data = await api.myCapsules(p, PAGE_SIZE);
      if (myId !== mineSeq) return;
      set((s) => ({
        mine: {
          ...s.mine,
          items: data.items,
          pagination: data.pagination,
          loading: false,
        },
      }));
    } catch (e) {
      if (myId !== mineSeq) return;
      set((s) => ({
        mine: {
          ...s.mine,
          loading: false,
          error: e instanceof Error ? e.message : "加载失败",
        },
      }));
    }
  },

  fetchFavorites: async (page) => {
    const p = page ?? get().favorites.page;
    const myId = ++favSeq;
    set((s) => ({
      favorites: { ...s.favorites, page: p, loading: true, error: null },
    }));
    try {
      const data = await api.myFavorites(p, PAGE_SIZE);
      if (myId !== favSeq) return;
      set((s) => ({
        favorites: {
          ...s.favorites,
          items: data.items,
          pagination: data.pagination,
          loading: false,
        },
      }));
    } catch (e) {
      if (myId !== favSeq) return;
      set((s) => ({
        favorites: {
          ...s.favorites,
          loading: false,
          error: e instanceof Error ? e.message : "加载失败",
        },
      }));
    }
  },

  deleteCapsule: async (id) => {
    await api.deleteMyCapsule(id);
    set((s) => ({
      mine: {
        ...s.mine,
        items: s.mine.items.filter((c) => c.id !== id),
      },
    }));
  },

  toggleFavorite: async (capsuleId, favoritedByMe) => {
    if (favoritedByMe) {
      await api.unfavorite(capsuleId);
      set((s) => ({
        favorites: {
          ...s.favorites,
          items: s.favorites.items.filter((c) => c.id !== capsuleId),
        },
      }));
      // 同步更新广场列表的收藏状态
      const plazaItem = usePlaza
        .getState()
        .items.find((i) => i.id === capsuleId);
      const newCount = Math.max(0, (plazaItem?.favoriteCount ?? 1) - 1);
      usePlaza.getState().patchFavorited(capsuleId, false, newCount);
      return { favoriteCount: newCount };
    } else {
      const result = await api.favorite(capsuleId);
      usePlaza
        .getState()
        .patchFavorited(capsuleId, true, result.favoriteCount);
      return { favoriteCount: result.favoriteCount };
    }
  },

  reset: () => {
    mineSeq++;
    favSeq++;
    set({ mine: emptySlice(), favorites: emptySlice() });
  },
}));

File: desktop/flutter/lib/stores/capsule.dart
// 我创建的 + 我收藏的：分页 / 删除 / 收藏切换（= React stores/capsule.ts）。
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../models/models.dart';
import 'plaza.dart';
import 'providers.dart';

const _pageSize = 15;

class ListSlice {
  final List<CapsuleListItem> items;
  final Pagination? pagination;
  final int page;
  final bool loading;
  final String? error;
  const ListSlice({
    this.items = const [],
    this.pagination,
    this.page = 1,
    this.loading = false,
    this.error,
  });

  ListSlice copyWith({
    List<CapsuleListItem>? items,
    Pagination? pagination,
    int? page,
    bool? loading,
    String? error,
    bool clearError = false,
  }) =>
      ListSlice(
        items: items ?? this.items,
        pagination: pagination ?? this.pagination,
        page: page ?? this.page,
        loading: loading ?? this.loading,
        error: clearError ? null : (error ?? this.error),
      );
}

class CapsuleListState {
  final ListSlice mine;
  final ListSlice favorites;
  const CapsuleListState({this.mine = const ListSlice(), this.favorites = const ListSlice()});
  CapsuleListState copyWith({ListSlice? mine, ListSlice? favorites}) =>
      CapsuleListState(mine: mine ?? this.mine, favorites: favorites ?? this.favorites);
}

class CapsuleNotifier extends Notifier<CapsuleListState> {
  int _mineSeq = 0;
  int _favSeq = 0;

  @override
  CapsuleListState build() => const CapsuleListState();

  Future<void> fetchMine([int? page]) async {
    final p = page ?? state.mine.page;
    final myId = ++_mineSeq;
    state = state.copyWith(mine: state.mine.copyWith(page: p, loading: true, clearError: true));
    try {
      final data = await ref.read(apiClientProvider).myCapsules(page: p, pageSize: _pageSize);
      if (myId != _mineSeq) return;
      state = state.copyWith(mine: state.mine.copyWith(items: data.items, pagination: data.pagination, loading: false));
    } catch (e) {
      if (myId != _mineSeq) return;
      state = state.copyWith(mine: state.mine.copyWith(loading: false, error: e.toString()));
    }
  }

  Future<void> fetchFavorites([int? page]) async {
    final p = page ?? state.favorites.page;
    final myId = ++_favSeq;
    state = state.copyWith(favorites: state.favorites.copyWith(page: p, loading: true, clearError: true));
    try {
      final data = await ref.read(apiClientProvider).myFavorites(page: p, pageSize: _pageSize);
      if (myId != _favSeq) return;
      state = state.copyWith(
          favorites: state.favorites.copyWith(items: data.items, pagination: data.pagination, loading: false));
    } catch (e) {
      if (myId != _favSeq) return;
      state = state.copyWith(favorites: state.favorites.copyWith(loading: false, error: e.toString()));
    }
  }

  Future<void> deleteCapsule(String id) async {
    await ref.read(apiClientProvider).deleteMyCapsule(id);
    state = state.copyWith(mine: state.mine.copyWith(items: state.mine.items.where((c) => c.id != id).toList()));
  }

  /// 切换收藏，返回最新收藏数；同步广场列表的收藏态。
  Future<int> toggleFavorite(String capsuleId, bool favoritedByMe) async {
    final api = ref.read(apiClientProvider);
    final plaza = ref.read(plazaProvider.notifier);
    if (favoritedByMe) {
      await api.unfavorite(capsuleId);
      state = state.copyWith(
          favorites: state.favorites.copyWith(items: state.favorites.items.where((c) => c.id != capsuleId).toList()));
      final matches = ref.read(plazaProvider).items.where((i) => i.id == capsuleId);
      final cur = matches.isEmpty ? null : matches.first;
      final newCount = ((cur?.favoriteCount ?? 1) - 1).clamp(0, 1 << 30);
      plaza.patchFavorited(capsuleId, false, newCount);
      return newCount;
    } else {
      final result = await api.favorite(capsuleId);
      plaza.patchFavorited(capsuleId, true, result.favoriteCount);
      return result.favoriteCount;
    }
  }

  void reset() {
    _mineSeq++;
    _favSeq++;
    state = const CapsuleListState();
  }
}

final capsuleProvider = NotifierProvider<CapsuleNotifier, CapsuleListState>(CapsuleNotifier.new);
